src/topology/perfect.lean

```diff
@@ -0,0 +1,279 @@
+ := by simp[acc_iff_cluster,cluster_pt_principal_iff_frequently]
+
+/-- If `x` is an accumulation point of `C` and `C ⊆ D`, then
+`x` is an accumulation point of `D. -/
+theorem acc_pt.mono {x : α} {C D : set α} (h : acc_pt x C) (hfg : C ≤ D) : acc_pt x D :=
+  ⟨ne_bot_of_le_ne_bot h.ne (inf_le_inf_left _ (monotone_principal hfg))⟩
+
+/-- If `x` is an accumulation point of `C` and `U` is a neighborhood of `x`,
+then `x` is an accumulation point of `U ∩ C`. -/
+theorem acc_pt.nhd_inter {x : α} {C U: set α} (h_acc : acc_pt x C) (hU : U ∈ 𝓝 x) :
+  acc_pt x (U ∩ C) :=
+begin
+  have : 𝓝[≠] x ≤ 𝓟 U,
+  { rw le_principal_iff,
+    exact mem_nhds_within_of_mem_nhds hU, },
+  rw [acc_pt, ← inf_principal, ← inf_assoc, inf_of_le_left this],
+  exact h_acc,
+end
+
+/-- A set `C` is preperfect if all of its points are accumulation points of itself.
+If `C` is nonempty, this is equivalent to the closure of `C` being perfect.
+See `preperfect_iff_closure_perfect`.-/
+--Note : This is my own term, feel free to suggest a better one :P
+def preperfect (C : set α) : Prop := ∀ x ∈ C, acc_pt x C
+
+/-- A set `C` is called perfect if it is closed and all of its
+points are accumulation points of itself.
+Note that we do not require `C` to be nonempty as is common,
+but see `perf_nonempty`.-/
+structure perfect (C : set α) : Prop :=
+  (closed : is_closed C)
+  (acc : preperfect C)
+
+/-- A set is called nonempty perfect if it is closed, nonempty, and
+all of its points are accumulation points of itself.-/
+structure perf_nonempty (C : set α) extends perfect C : Prop :=
+  (nonempty : C.nonempty)
+
+theorem preperfect_iff_nhds {C : set α} : preperfect C ↔ ∀ x ∈ C, ∀ U ∈ 𝓝 x, ∃ y ∈ U ∩ C, y ≠ x
+  := by simp only[preperfect, acc_pt_iff_nhds]
+
+/-- The intersection of a preperfect set and an open set is preperfect-/
+theorem preperfect.open_inter {C U : set α} (hC : preperfect C) (hU : is_open U) :
+  preperfect (U ∩ C) :=
+begin
+  rintros x ⟨xU, xC⟩,
+  apply (hC _ xC).nhd_inter,
+  exact hU.mem_nhds xU,
+end
+
+/-- The closure of a preperfect set is perfect.
+For a converse, see `preperfect_iff_perfect_closure`-/
+theorem preperfect.perfect_closure {C : set α} (hC : preperfect C) :
+  perfect (closure C) :=
+begin
+  split, {apply is_closed_closure},
+  intros x hx,
+  by_cases h : x ∈ C; apply acc_pt.mono _ subset_closure,
+  { exact hC _ h },
+  have : {x}ᶜ ∩ C = C := by simp[h],
+  rw [acc_pt, nhds_within, inf_assoc, inf_principal, this],
+  rw [closure_eq_cluster_pts] at hx,
+  exact hx,
+end
+
+--perhaps this belongs in a different file?
+theorem frequently_nhds_nhds {x : α} {p : α → α → Prop} :
+  (∃ᶠ y in 𝓝 x, ∃ᶠ z in 𝓝 y, p x z) → (∃ᶠ z in 𝓝 x, p x z) :=
+begin
+  rw [(nhds_basis_opens _).frequently_iff, (nhds_basis_opens _).frequently_iff],
+  rintros h U ⟨xU,Uop⟩,
+  rcases h U ⟨xU,Uop⟩ with ⟨y,yU,hy⟩,
+  rw (nhds_basis_opens _).frequently_iff at hy,
+  rcases hy U ⟨yU,Uop⟩ with ⟨z,zU,hz⟩,
+  use [z,zU,hz],
+end
```

Comment:
This should be near `eventually_eventually_nhds`. Note that it is even an `iff`, and that you can prove it by reducing it to the `eventually` case:
```lean
begin
  rw ← not_iff_not,
  simp_rw not_frequently,
  exact eventually_eventually_nhds,
end
```

Reply:
done. Thanks for the suggestion! I wasn't able to find eventually_eventually_nhds before.